src/ext/java/org/opentripplanner/ext/fares/model/FareLegRule.java

```diff
@@ -1,22 +1,36 @@
 package org.opentripplanner.ext.fares.model;
 
+import java.util.Collection;
+import java.util.List;
+import java.util.Objects;
 import javax.annotation.Nonnull;
 import javax.annotation.Nullable;
+import org.opentripplanner.framework.collection.ListUtils;
 import org.opentripplanner.model.fare.FareProduct;
+import org.opentripplanner.transit.model.framework.FeedScopedId;
 
 public record FareLegRule(
-  @Nullable String legGroupId,
+  @Nonnull FeedScopedId id,
+  @Nullable FeedScopedId legGroupId,
   @Nullable String networkId,
   @Nullable String fromAreaId,
   @Nullable String toAreaId,
   @Nullable FareDistance fareDistance,
-  @Nonnull FareProduct fareProduct
+  @Nonnull Collection<FareProduct> fareProducts
 ) {
+  public FareLegRule {
+    Objects.requireNonNull(id);
+    Objects.requireNonNull(fareProducts);
```

Comment:
Can you add a check that `fareProducts` is not empty and that it doesn't contain null?

Reply:
This probably means that you will have to adjust the test, which is a good thing.